File: rtc_code_editor/app/channels/doc_channel.rb
class DocChannel < ApplicationCable::Channel
  def subscribed
    stream_from "doc_channel_#{params[:document_id]}"
    @user = connection.current_user
    if params[:editing]
      send_join_request
      announce_arrival
    end

    send_initial_state
  end

  def receive(data)
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", data)
  end

  def unsubscribed
    if params[:editing]
      remove_location
      announce_departure
    end
  end

  private

  def remove_location
    signal = { 
      exit: true, 
      senderId: connection.current_user.id, 
      senderName: connection.current_user.username 
    }
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", signal)
  end

  def send_join_request
    signal = { join: true, senderId: connection.current_user.id }
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", signal)
  end

  def send_initial_state
    syncRequest = { sender_id: connection.current_user.id, sync: true }
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", syncRequest)
  end

  def announce_arrival
    puts "@user: #{@user}"
    user_data = @user.attributes.slice('id', 'avatar_url', 'username', 'email')
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", {arrival: user_data})
  end

  def announce_departure
    ActionCable.server.broadcast("doc_channel_#{params[:document_id]}", {departure: @user.id})
  end
end

File: rtc_code_editor/app/javascript/channels/doc_channel.js
import consumer from "./consumer"

   const connectToDoc = (docId, editing, callbacks, userId) => {
    return consumer.subscriptions.create({
      channel: "DocChannel", 
      document_id: docId,
      editing
    }, {
      connected() {
        if (editing) callbacks.connect();
      },

      received(data) {
        if (data.join || data.offer || data.iceCandidate || data.answer) {
          if (data.senderId === userId || !editing) return;
          callbacks.initConnection(data);
          callbacks.assignColor(data);
          return
        }

        if (data.exit) {
          callbacks.assignColor(data);
        }

        if (editing && data.revocation && data.revocation.recipient_id === userId) {
          callbacks.ejectUser();
        }
        
        if (data.saved_state) {
          callbacks.save(data)
          return;
        }
        
        if (data.sync) {
          callbacks.sendState(data);
          return;
        }

        if(data.currentState) {
          callbacks.syncState(data);
          return;
        }
        
        if (data.document) {
          callbacks.initialize(data)
          return;
        }
        
        // if (data.editors) {
        //   callbacks.editorList(data);
        //   return;
        // }

        if (data.arrival || data.departure) {
          callbacks.editorList(data);
        }

        const update = data.backup;
        if (!update) return;
        
        if (update.changeData) { 
          callbacks.edit(data);
        } 

        if (update.position) {
          callbacks.cursor(data);
        }
      },

      resubscribe() {
        this.perform("subscribed");
      }
    });
  }


  export default connectToDoc;